sync3/handler/connstate_live.go

```diff
@@ -284,22 +284,45 @@ func (s *connStateLive) processUpdatesForSubscriptions(ctx context.Context, buil
 
 // this function does any updates which apply to the connection, regardless of which lists/subs exist.
 func (s *connStateLive) processGlobalUpdates(ctx context.Context, builder *RoomsBuilder, up caches.Update) (delta sync3.RoomDelta) {
-	rup, ok := up.(caches.RoomUpdate)
-	if ok {
-		delta = s.lists.SetRoom(sync3.RoomConnMetadata{
-			RoomMetadata: *rup.GlobalRoomMetadata(),
-			UserRoomData: *rup.UserRoomMetadata(),
-		})
-	}
+	var roomActivityTimestamp uint64
 
 	roomEventUpdate, ok := up.(*caches.RoomEventUpdate)
 	if ok {
-		// TODO: we should do this check before lists.SetRoom
 		if roomEventUpdate.EventData.LatestPos <= s.loadPosition {
 			return // if this update is in the past then ignore it
 		}
 		s.loadPosition = roomEventUpdate.EventData.LatestPos
+
+		// bump this room if this connection cares about this event type.
+		if s.muxedReq.BumpEventTypes != nil && len(s.muxedReq.BumpEventTypes) > 0 {
+			for _, eventType := range s.muxedReq.BumpEventTypes {
+				if eventType == roomEventUpdate.EventData.EventType {
+					roomActivityTimestamp = roomEventUpdate.EventData.Timestamp
+					break
+				}
+			}
+		} else {
+			roomActivityTimestamp = roomEventUpdate.EventData.Timestamp
+		}
 	}
+
+	inviteEventUpdate, ok := up.(*caches.InviteUpdate)
+	if ok {
+		// Even if we have ignored membership events, we probably want to know when
+		// we've been invited to a room.
+		roomActivityTimestamp = inviteEventUpdate.InviteData.LastMessageTimestamp
```

Comment:
I feel uneasy with this because it feels like we're going to inevitably add in a lot of exceptions. I wonder if a better approach would be to just modify the `LastMessageTimestamp` directly, instead of sometimes copying over the value, sometimes sending 0 to mean no change, etc. This would also mean the tests don't need to duplicate the timestamp, and should make it much less likely to break everything since you're just very situationally (if `bump_event_types` is set) modifying the `LastMessageTimestamp` rather than fundamentally changing the sort logic to use `LastActivityTimestamp`.

We reliably return copies of `RoomMetadata` so you can edit it freely and not pollute other connections, see https://github.com/matrix-org/sliding-sync/blob/main/sync3/caches/global.go#L79

We have precedent for modifying this struct when calculating the room name, as we modify the `Heroes` slice to remove the user ID of the caller, see https://github.com/matrix-org/sliding-sync/blob/main/sync3/handler/connstate.go#L94 and https://github.com/matrix-org/sliding-sync/blob/main/internal/roomname.go#L65

Reply:
Moreover it seems like that heroes modification ought to be done in `UserRoomData` rather than `RoomMetadata`?